### PR TITLE
feature(report): Templates V2 - Elastic abstaction

### PR DESCRIPTION
**Pull request by oxypomme**
# Report

> [!IMPORTANT]  
> Old templates (template v1) support is dropped, please use `ezreeport-admin migrate <data-folder>` to transform your templates to the new format

## Features

- Template V2
 - Simplified aggregation management and other figures params
 - Template V1 support is dropped, please use `ezreeport-admin migrate <data-folder>` to transform your templates to the new format

## Bug fixs

- More details with errors while rendering 
- Fixed various timezone issues

# ezreeport-admin

Added migration

### CODE REVIEW

File: services/report/src/models/reports/fetch/index.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  recurrence: Type.Enum(Recurrence),
+  period: Type.Object({
+    start: Type.Integer(),
+    end: Type.Integer(),
+  }),
+  auth: Type.Object({
+    username: Type.String({ minLength: 1 }),
+  }),
+  dateField: Type.String({ minLength: 1 }),
+  index: Type.Optional(Type.String({ minLength: 1 })),
```
Comment (felixleo22) [inline services/report/src/models/reports/fetch/index.ts]: Not optional

File: services/report/src/models/reports/fetch/index.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    .filter((f) => f.type !== 'md');
+
+  const requests: ElasticTypes.MsearchMultisearchBody[] = figures.map((f) => {
+    const query = prepareEsQuery(
+      [options.filters, f.filters].filter((x) => !!x).flat(), // Merge filters
+      options.dateField,
+      options.period,
+    );
+
+    // TODO: add cause in error
```
Comment (felixleo22) [inline services/report/src/models/reports/fetch/index.ts]: Bro ?

File: services/report/src/models/reports/fetch/results.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        if (!('value' in aggregation) || typeof aggregation.value !== 'number') {
+          throw new Error('Aggregation value is missing for a metric figure');
+        }
+
+        return {
+          key: `${label.text}`,
+          value: aggregation.value ?? 0,
+        };
+      },
+    ).filter((result) => !!result);
```
Comment (felixleo22) [inline services/report/src/models/reports/fetch/results.ts]: Remove .filter((result) => !!result);

File: services/report/src/models/reports/fetch/results.ts [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+};
+
+export default function handleEsResponse(
+  response: EsResponse,
+  figure: FigureType,
+  errorCause: Record<string, unknown>,
+): FetchResultItem[] {
+  checkEsErrors(response, errorCause);
+
+  if (!('aggregations' in response) || !response.aggregations) {
```
Comment (felixleo22) [inline services/report/src/models/reports/fetch/results.ts]: Move it in checkEsErrors

File: services/report/src/app.ts [hunk 5]
```diff
@@ -18,6 +19,9 @@ const { port, allowedOrigins: rawOrigins } = config;
 const start = async () => {
   // Create Fastify instance
   const fastify = Fastify({
+    ajv: {
```
Comment (felixleo22) [inline services/report/src/app.ts]: Check https://github.com/ezpaarse-project/readlabs/blob/main/services/api/src/app.ts#L54

File: services/report/src/models/reports/render.ts [hunk 6]
```diff
@@ -27,6 +27,7 @@ import { Type, type Static, assertIsSchema } from '~/lib/typebox';
 
 import type { FigureType } from '~/models/figures';
 import { Layout } from '~/models/layouts';
+import { FetchResultItem } from './fetch/results';
```
Comment (felixleo22) [inline services/report/src/models/reports/render.ts]: import type

File: services/report/src/lib/pdf/metrics.ts [hunk 7]
```diff
@@ -105,14 +100,45 @@ const formatNumber = (
...
+          break;
+
+        default:
+          value = `${value}`;
+          break;
+      }
+    }
+  } catch (error) {
+    if (!(error instanceof Error)) {
+      throw new Error(`An unexpected error occurred wile formatting "${label.text}" ("${value}"): ${error}`);
```
Comment (felixleo22) [inline services/report/src/lib/pdf/metrics.ts]: wile

**Review by nojhamster — APPROVED** [review]

**Pull request merged**